design/tagging-API-change.md

```diff
@@ -0,0 +1,73 @@
+# Tagging API Change
+
+CDK support tagging and can cascade tags to all its taggable children (see [here](https://docs.aws.amazon.com/cdk/latest/guide/tagging.html)). The current CDK tagging API is shown below:
+
+``` ts
+myConstruct.node.applyAspect(new Tag('key', 'value'));
+
+myConstruct.node.applyAspect(new RemoveTag('key', 'value'));
+```
+
+As we can see, the current tagging API is not nice and grammatically verbose for using, since there is no reason to expose `node` to users and `applyAspect` does not indicate anything towards tags, which leaves room for improvement. Also, users need to create two objects to add tag and remove tag which causes confusion to some degree.
+
+## General approach
+
+For the tagging behavior part, we propose using just one entry point `Tag` for the new tagging API:
+
+``` ts
+Tag.add(myConstruct, 'key', 'value');
+
+Tag.remove(myConstruct, 'key');
+```
+
+## Code changes
+
+Given the above, we should make the following changes:
+  1. Add two methods `add` and `remove` to `Tag` class, which calls `applyAspect`to add tags or remove tags.
+
+# Part1: Change CDK Tagging API
+
+Implementation for the new tagging API is shown below:
+
+``` ts
+/**
+ * The Tag Aspect will handle adding a tag to this node and cascading tags to children
+ */
+export class Tag extends TagBase {
+
+  /**
+   * add tags to the node of a construct and all its the taggable children
+   */
+  public static add(scope: Construct, key: string, value: string, props: TagProps = {}) {
```

Comment:
since tags are essentially maps, why not something like this:

```ts
Tag.add(scope, { sfoo: 'bar' })
```

This will allow people to add multiple tags like this:

```ts
Tag.add(scope, {
  tag1: 'value1',
  tag2: 'value2'
});
```

Reply:
But tag also supports other properties like priority. See [here](https://docs.aws.amazon.com/cdk/latest/guide/tagging.html).

Reply:
okay, let's just leave it with `Tag.add(scope, key, value[, options)`